### PR TITLE
Adds Async Shader support

### PR DESCRIPTION
**Pull request by Kuranes**
- Adds shaderSync  url options
- only activates on PBR shaders

### CODE REVIEW

File: sources/osgShader/ShaderGenerator.js [hunk 1]
```diff
@@ -295,6 +366,15 @@ ShaderGenerator.prototype = {
             program.generated = true;
 
             this._cache[hash] = program;
+
+            // first async compilation
+            program.apply(state, !synchronizedCompilation);
```
Comment (cedricpinson) [inline sources/osgShader/ShaderGenerator.js]: not sure apply need an extra argument if program has extra data like the _placeHolder or asyncCompilation flag

File: sources/osg/Program.js [hunk 2]
```diff
@@ -412,11 +431,22 @@ utils.createPrototypeStateAttribute(
...
+                    return;
+                }
+
+                this.compile();
 
-                this._compileClean = this.compile();
-                this.getCompilationResult(this._compileClean, this._gl);
-                state.applyProgram(this._program);
-                this._dirty = false;
+                if (!async || this._shaderName.indexOf('PBR') === -1) {
```
[the remark this answers is not in the document]
Comment (cedricpinson) [reply sources/osg/Program.js]: why asyncing only pbr ?

[the remark this answers is not in the document]
Comment (Kuranes) [reply sources/osg/Program.js]: other shader might give problems: (ie materialpacker where you need direct results on one frame)

File: sources/osgShader/ShaderGenerator.js [hunk 3]
```diff
@@ -237,6 +241,27 @@ ShaderGenerator.prototype = {
         CompilerShader._validTextureAttributeTypeMemberCache = cache;
     },
 
+    _createWaitingProgram: function(state, attributes, textureAttributes) {
+        var vsDefault =
+            'attribute vec3 Vertex;uniform mat4 uModelViewMatrix;uniform mat4 uProjectionMatrix;void main(void) {  gl_Position = uProjectionMatrix * (uModelViewMatrix * vec4(Vertex, 1.0));}\n';
+
+        var fsDefault =
```
[the remark this answers is not in the document]
Comment (Kuranes) [reply sources/osgShader/ShaderGenerator.js]: will try

**Pull request merged**